QA Review UI

SuaYoo opened the pull request:
Partially addresses https://github.com/webrecorder/browsertrix-cloud/issues/1477

<!-- Fixes #issue_number -->

### Changes

- #1497
- #1496
- #1499
- #1589
- #1508

SuaYoo committed:
get latest qa run

SuaYoo committed:
Add QA screenshot comparison tool (#1552)

Adds initial screenshot comparison tool for testing RWP embeds.

SuaYoo committed:
Add QA screenshot comparison tool (#1552)

Adds initial screenshot comparison tool for testing RWP embeds.

SuaYoo committed:
enable qa tab

emma-sg committed:
QA page list sidebar (#1546)

- Adds a Grouped List helper function:
[`frontend/src/components/utils/grouped-list.ts`](https://github.com/webrecorder/browsertrix-cloud/pull/1546/files#diff-7cdd94b36ceca715ce50d68524a31386741561bccdbe1760e1dfae8157f5495e)
- Adds a `WeakRef`-based function cache for memoizing function calls,
along with tests for it:
[`frontend/src/utils/weakCache.ts`](https://github.com/webrecorder/browsertrix-cloud/pull/1546/files#diff-a3164accbef418eedb596ccbb96cf0dee4d4c594fef28a80f02124352875109a)
- Adds a page list, along with a bunch of functional helpers:
[`frontend/src/features/qa/page-list/page-list.ts`](https://github.com/webrecorder/browsertrix-cloud/pull/1546/files#diff-59fb183decceed95276697a1c274d210b4d4d689e4b5d6eed713b331dfb5b804)
- Adds `clsx` for combining classes
- Adds a [`TailwindComponent` template for
VSCode](https://github.com/webrecorder/browsertrix-cloud/pull/1546/files#diff-64ebc963103ab4a0b8781ad0173ea0de10c4112303531db7c592c6773e25d626)
- Reverts changes to the `<btrix-badge>` component
- Improves [ItemListControl
types](https://github.com/webrecorder/browsertrix-cloud/pull/1546/files#diff-e2e161f4ce6410ed4a5605469aa2260fabdd384e4fcfd4e2978e85d043912579)
- Adds an animation helper from Shoelace:
[`frontend/src/features/qa/page-list/ui/animate.ts`](https://github.com/webrecorder/browsertrix-cloud/pull/1546/files#diff-5e74f8833205ed5700da8ba2ea81d8aff110a93add8bd19c35ff5cd579f71a4d)
- Adds and enables `@tailwindcss/container-queries` for future use
- Updates `prettier-plugin-tailwindcss` and ensures that the VSCode
Prettier extension uses the correct prettier config & plugins

SuaYoo committed:
try grouping pages

SuaYoo committed:
back to single list with filtering and qa data

SuaYoo committed:
update number formatting

SuaYoo committed:
update sorting

SuaYoo committed:
show active page qa analysis

SuaYoo committed:
add back scroll into view

SuaYoo committed:
add tooltip and update layout

SuaYoo committed:
update layout

SuaYoo committed:
add qa run dropdown

SuaYoo committed:
add tab group

SuaYoo committed:
enable select qa run

SuaYoo committed:
fix qa run dropdown

ikreymer committed:
Frontend QA: Add Text + Resources Tabs + Optimizations (#1641)

Optimize loading: create a proxy iframe for loading from service worker,
to be able to fetch() directly from serviceworker:
- fetch to blob and use blob URL for `<img>`
- fetch to text for text tab
- fetch json for resources

Text Tab: Show text content in each pane, use diff library for text diff

Resources: Aggregate resources JSON by 'type', also group into good
(status < 400) /bad (status >= 400)

Minor styling tweaks: split nav bar into two rows, add height to replay.
Only load url in RWP when actually showing replay to avoid unnecessary
background loading.

Update data when QA Item is toggled

---------
Co-authored-by: sua yoo <sua@suayoo.com>

SuaYoo committed:
submit qa review

Shrinks99 commented:
Will take a deeper look tomorrow, initial thought: This is starting to really come together!!

~~I liked the filled next page button?  There's a whole lot of buttons to press on this page, and I think giving the "next" one more hierarchy importance probably isn't a bad move?  I've given _Finish Review_ the `primary` flag in https://github.com/webrecorder/browsertrix-cloud/pull/1645, generally like how this looks but open to opinions?~~

EDIT: Lets save button colours for later, don't wanna get hung up on them here!

ikreymer committed:
proxy iframe: use '/placeholder' URL to avoid loading RWP again

ikreymer committed:
add crawlDataRegistered / qaDataRegistered bools to track if the QA / crawl item endpoint
is registered in RWP SW, even if there is no data.
if not registered, triggers injecting RWP and awaiting message

Shrinks99 committed:
Minor UI additions for QA (#1645)

### Changes
- Changes second level status icon to slash-square (app wide)
- Updates finish review button
- Changes to success state, hopefully better exemplifies that this is
the final action compared to "primary" which is used for things we want
to highlight?
- Adds icon
- Adds microscope icon for review button

ikreymer committed:
fix typo, add null check for event.source (for safari)
disable devServer compression for firefox

ikreymer commented:
Added some additional fixes, and I think this may be ready to merge!

ikreymer commented on .github/workflows/frontend-build-check.yaml:
```diff
@@ -37,7 +37,7 @@ jobs:
         # TODO Reenable when https://github.com/webrecorder/browsertrix-cloud/issues/1618 is addressed
         # run: yarn format:check
-        run: yarn prettier --list-different .
+        run: echo "yarn format:check disabled"
```
@emma-sg @SuaYoo assume this is intentional, right?

emma-sg replied:
Unsure... I think there was a tailwind class sorting issue where certain classes would be sorted differently in CI vs locally, but I'm not sure that's reason enough to totally disable this check

ikreymer replied:
Hm, maybe we revisit it later, there's already a todo there.

ikreymer commented on frontend/frontend.conf.template:
```diff
@@ -45,6 +45,7 @@ server {
     # serve replay service worker, RWP_BASE_URL set in Dockerfile
     location /replay/sw.js {
       add_header Content-Type application/javascript;
+      add_header Service-Worker-Allowed /;
```
Hm, not sure if we need this...

ikreymer committed:
frontend: make /replay/ path serve 404 with error to more easily catch fallthroughs, remove unneeded Server-Worker-Allowed header, in both nginx and devServer

ikreymer committed:
Merge branch 'main' into frontend-qa-analysis

ikreymer committed:
fix nginx default type for /replay/

ikreymer committed:
flip text compares

Shrinks99 commented on frontend/src/pages/org/archived-item-qa/ui/text.ts:
```diff
@@ -0,0 +1,105 @@
+
+      return html`
+        <div
+          class=${tw`flex-1 overflow-hidden whitespace-pre-line rounded-lg border-dashed p-4 first-of-type:border-r`}
```
I found the text diff much easier to parse as a large block.  While not exactly easy to _read_, I found it was quicker to see related parts of the page that were differences as a large block of text instead of multiple lines where the relation between those elements was less clear to me.

Also less scrolling!  Over many pages this will compound into lots of time saved!

```suggestion
          class=${tw`flex-1 overflow-hidden rounded-lg border-dashed p-4 first-of-type:border-r`}
```

ikreymer replied:
Let's try it, as this may look better with the diff tool we're using

Shrinks99 commented on frontend/src/pages/org/archived-item-qa/ui/text.ts:
```diff
@@ -0,0 +1,105 @@
+          })}
+        </div>
+        <div
+          class=${tw`flex-1 overflow-hidden whitespace-pre-line rounded-lg border-dashed p-4 first-of-type:border-r`}
```
```suggestion
          class=${tw`flex-1 overflow-hidden rounded-lg border-dashed p-4 first-of-type:border-r`}
```

Shrinks99 commented on frontend/src/pages/org/archived-item-qa/ui/text.ts:
```diff
@@ -0,0 +1,105 @@
+            return html`
+              <span
+                class=${part.added
+                  ? removedText
```
It appears the wrong text is ending up in each box!  See "Archived page not found" in the crawl tab!

<img width="814" alt="Screenshot 2024-04-03 at 9 05 30 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5672810/ec1f9f40-3c06-4d2c-b7b7-4e5a8a12df99">

ikreymer replied:
This has been fixed now

emma-sg commented on frontend/src/components/ui/tab-group/tab.ts:
```diff
@@ -0,0 +1,67 @@
+
+import "@/components/ui/navigation/navigation-button";
+
+import { NavigationButton } from "@/components/ui/navigation/navigation-button";
```
Does `@/components/ui/navigation/navigation-button` need to be imported twice?

emma-sg commented on frontend/src/features/qa/page-list/ui/page-details.ts:
```diff
@@ -0,0 +1,108 @@
+          tw`mr-2 flex-none`,
+        )}
+        <span class="inline-block">
+          ${page.qa.screenshotMatch != null
```
Is `screenshotMatch` ever going to be null? I remember it could be at one point, did we perform a migration to ensure it never is? If not, let's update the type here.

ikreymer replied:
Yes, we should assume that this may be optional, as all of the QA data may not be available for whatever reason (screenshot, text, resources)

emma-sg commented on frontend/src/features/qa/page-list/helpers/issueCounts.ts:
```diff
@@ -0,0 +1,48 @@
+    severityFromMatch(page.qa.screenshotMatch),
+    severityFromMatch(page.qa.textMatch),
+    severityFromResourceCounts(
+      page.qa.resourceCounts?.crawlBad,
+      page.qa.resourceCounts?.crawlGood,
+    ),
+    severityFromResourceCounts(
+      page.qa.resourceCounts?.replayBad,
+      page.qa.resourceCounts?.replayGood,
```
Similarly, is it guaranteed that we'll have values for props of `resourceCounts`? If not, let's update the type to reflect this.

ikreymer replied:
`resourcesCounts` is not guaranteed, but entries within, if it exists, should be. Will update!

emma-sg commented on frontend/src/pages/org/archived-item-detail/archived-item-detail.ts:
```diff
@@ -123,17 +158,35 @@ export class CrawlDetail extends LiteElement {
+
+      this.fetchTabData();
+    } else {
+      if (changedProperties.has("activeTab") && this.activeTab) {
```
is `this.activeTab` ever going to be falsey?

ikreymer replied:
Not sure, but maybe revisit later for this one?

ikreymer replied:
Want to err on side of caution for now.

emma-sg commented on frontend/src/pages/org/archived-item-qa/archived-item-qa.ts:
```diff
@@ -0,0 +1,1012 @@
+        const text = await resp.text();
+        return { text };
+      } else if (tab === "resources") {
+        const json = await resp.json();
```
Do we know the type of this?

ikreymer replied:
Added!

emma-sg commented on frontend/src/pages/org/archived-item-qa/archived-item-qa.ts:
```diff
@@ -0,0 +1,1012 @@
+
+        for (const entry of Object.values(json.urls)) {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const { type: resType_, status } = entry as any;
```
It'd be ideal to avoid `any`s, if possible

emma-sg committed:
fix a bunch of linting issues

ikreymer committed:
type fixes from code review

ikreymer committed:
Remove whitespace-pre-line from text diff (from code review)

Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

ikreymer committed:
backend patch: ensure new_status is set to None

emma-sg committed:
allow a couple more types to be undefined
